skill-tree/src/main/java/com/RDS/skilltree/services/EndorsementServiceImplementation.java

```diff
@@ -1,39 +1,65 @@
 package com.RDS.skilltree.services;
 
-import com.RDS.skilltree.User.UserModel;
-import com.RDS.skilltree.User.UserRepository;
-import com.RDS.skilltree.User.UserSkillStatusEnum;
-import com.RDS.skilltree.User.UserSkillsModel;
+import com.RDS.skilltree.dtos.RdsGetUserDetailsResDto;
 import com.RDS.skilltree.exceptions.EndorsementNotFoundException;
 import com.RDS.skilltree.exceptions.SelfEndorsementNotAllowedException;
 import com.RDS.skilltree.exceptions.SkillNotFoundException;
-import com.RDS.skilltree.exceptions.UserNotFoundException;
 import com.RDS.skilltree.models.Endorsement;
+import com.RDS.skilltree.models.JwtUser;
 import com.RDS.skilltree.models.Skill;
+import com.RDS.skilltree.models.UserSkills;
 import com.RDS.skilltree.repositories.EndorsementRepository;
 import com.RDS.skilltree.repositories.SkillRepository;
 import com.RDS.skilltree.repositories.UserSkillRepository;
+import com.RDS.skilltree.services.external.RdsService;
 import com.RDS.skilltree.viewmodels.CreateEndorsementViewModel;
 import com.RDS.skilltree.viewmodels.EndorsementViewModel;
 import com.RDS.skilltree.viewmodels.UpdateEndorsementViewModel;
-import java.util.List;
-import java.util.Objects;
-import java.util.Optional;
+import com.RDS.skilltree.viewmodels.UserViewModel;
+import java.util.*;
 import lombok.RequiredArgsConstructor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.core.context.SecurityContextHolder;
 import org.springframework.stereotype.Service;
 
 @Service
 @RequiredArgsConstructor
 public class EndorsementServiceImplementation implements EndorsementService {
-    private final UserRepository userRepository;
+    private static final Logger log = LoggerFactory.getLogger(EndorsementServiceImplementation.class);
+    private final RdsService rdsService;
     private final SkillRepository skillRepository;
     private final EndorsementRepository endorsementRepository;
     private final UserSkillRepository userSkillRepository;
 
     @Override
     public List<EndorsementViewModel> getAllEndorsementsBySkillId(Integer skillId) {
```

Comment:
here is the possibility of redundant API calls. The code makes two separate calls to rdsService.getUserDetails for each endorsement, one for the endorseId and another for the endorserId.

There are two ways this could be optimized:

Combining calls based on userIds:

You can potentially combine the API calls if both endorseId and endorserId are sent in a single request to rdsService.getUserDetails. If the service supports fetching multiple users in one call, you can create a list containing both endorseId and endorserId, and then use that list to fetch user details in one go. This would eliminate redundant calls for users already fetched.

Checking for duplicates before fetching:
Inside the loop, you can check if the user details for a particular endorseId or endorserId already exist in the userDetails map before calling the rdsService.getUserDetails. This would avoid making unnecessary calls for users already retrieved.

```
public List<EndorsementViewModel> getAllEndorsementsBySkillId(Integer skillId) {
    List<Endorsement> endorsements = endorsementRepository.findBySkillId(skillId);

    // store all users data that are a part of this request
    Map<String, UserViewModel> userDetails = new HashMap<>();

    return endorsements.stream()
        .map(
            endorsement -> {
                String endorseId = endorsement.getEndorseId();
                String endorserId = endorsement.getEndorserId();

                if (!userDetails.containsKey(endorseId)) {
                    RdsGetUserDetailsResDto endorseDetails = rdsService.getUserDetails(endorseId);
                    userDetails.put(endorseId, UserViewModel.toViewModel(endorseDetails.getUser()));
                }

                if (!userDetails.containsKey(endorserId)) {
                    RdsGetUserDetailsResDto endorserDetails = rdsService.getUserDetails(endorserId);
                    userDetails.put(endorserId, UserViewModel.toViewModel(endorseDetails.getUser()));
                }

                return EndorsementViewModel.toViewModel(
                    endorsement, userDetails.get(endorseId), userDetails.get(endorserId));
            })
        .toList();
}

```